### ISSUE TITLE
Improved YAML format

### ISSUE THREAD

**Issue by mildred** [purpleidea/mgmt#141]
Until we have the perfect DSL, I would suggest to improve the YAML format, and I have one idea for that:

```
version: 2
graph: mygraph
resources:
  - name: file1
    autoedge: true
    retry: -1
    # the presence of the file key denotes a file resource
    file:
      path: /tmp/mgmt/file1
      content: |
        Content of my file
      owner: root
  - name: docker
    # The presence of a svc key denotes a svc resource
    svc:
      state: running
      startup: enabled
    after: # list of resource it depends on, this create edges
      - file file1 # format: <resource kind><space><resource name>
    before: # list of resources that depends on this
      - noop checkpoint
  - name: checkpoint
    noop: true
edges:
  # additional edges just as before
```

The implementation of this (which I started) is even simpler because it doesn't need the reflect package in `yamlgraph/gconfig.go` in function `NewGraphFromConfig()`. The `resources` key already contains the list.

**Comment by roidelapluie** [comment]
I would rather propose:

```
  - name: file1
    autoedge: true
    retry: -1
    kind: file
    params:
      path: /tmp/mgmt/file1
      content: |
        Content of my file
      owner: root
    metaparams:
```

**Comment by roidelapluie** [comment]
But this is great! Even if I would rater prefer us to focus on the DSL instead of this implementation point.

**Comment by mildred** [comment]
There is just one catch with this format: a resource name cannot be one of `name`, `before`, `after`, or one of the meta parameters (`autoedge`, `autogroup`, `retry`, `delay`, `poll`, `limit`, `burst`). Except for `noop` which can be a special case. if this restriction is not wanted, the syntax suggested by @roidelapluie is a nice alternative.

**Comment by mildred** [comment]
I admit I'm biased towards yaml. This is one thing that I love in ansible. Clear syntax, easy to learn, machine processable with any language...

**Comment by roidelapluie** [comment]
I like yaml too, and as it is our interface with mgmt this makes sense to improve it.

**Comment by purpleidea** [comment]
I'm not explicitly against changing the YAML format, but as I've mentioned, it is only a crutch for me until we have the language, and so I don't plan to spend any cycles on this. If someone feels *really* strongly about this, and the new format doesn't have any limitations, then patches welcome I suppose. But I'm going to close this issue.

**Comment by roidelapluie** [comment]
Please @purpleidea stop closing issues all the time.

**Comment by purpleidea** [comment]
@roidelapluie Opening infinite numbers of issues without fixes on the near horizon isn't productive to me, since it just adds to the noise. If you feel really strongly about an issue that I've closed, LMK which and I can reopen. This particular issue is WONTFIX as far as I'm concerned. I see there's a patch. I'l have a look

**Comment by ffrank** [comment]
You cats, I feel that this one is a little misguided. Sorry ><'

Here's my POV: mgmt does *not* expose a YAML-powered interface for the user to construct a graph. What's happening is that the tool exposes its internal data structure so that YAML can be used (arbitrarily) as a format to cobble up such internal representations directly (in lieu of an actual interface that deserves the name).

As such, making "the YAML" more user-friendly is way besides the point. These are internal data structures close to the heart of the application. Focus should be imho
 * code maintainability (i.e. mgmt's golang code)
 * performance

So, with respect for @mildred's love for YAML-based interfaces, and the tremendous amount of work you've poured into #143 already, I'm honestly not in favor of adding this overhead to mgmt.

**Comment by roidelapluie** [comment]
@ffrank Agree, but I see this more as a "new parser" that could approach a basic "DSL" in yaml ..

**Comment by mildred** [comment]
I started using mgmt, including with the new syntax I offered in #143 and I admit that a DSL would be nice to have at this point. Improving the yaml format is not enough.

A DSL could:

- improve the way the resources are grouped together. You might want to have 3 resources for a particular purposes separate from 3 others.
- multiple files and includes could be nice to have, in particular wildcard or directory includes where you just have to drop a file in a directory to have it processed
- and the ton of things a DSL could do. Variables, functions, ...

Still, I think the YAML format could be improved. A subset of #143 could be nice to have (the subset that doesn't change the current format). I'll comment more on this on #143.

**Comment by ffrank** [comment]
Another important thing about the DSL: It will be part of mgmt's larger design architecture. mgmt will not be able to really use facter, ohai, or Ansible's setup module. mgmt's analog of facts will have to be continuous sources of data that can raise events. Any piece of mgmt DSL code will be able to affect the graph (profoundly) on occurrences of all kinds of events.

In order to make this manageable, we will need quite a high-level language. Exposing the immediate graph structure to the user like now will probably not be an option anymore.
